src/StreamJsonRpc/StandardCancellationStrategy.cs

```diff
@@ -0,0 +1,144 @@
+    using Microsoft.VisualStudio.Threading;
+    using StreamJsonRpc.Protocol;
+
+    internal class StandardCancellationStrategy : ICancellationStrategy
+    {
+        /// <summary>
+        /// The JSON-RPC method name used to send/receive cancellation requests.
+        /// </summary>
+        private const string CancelRequestSpecialMethod = "$/cancelRequest";
+
+        /// <summary>
+        /// The <see cref="MethodInfo"/> for the <see cref="CancelInboundRequest(RequestId)"/> method.
+        /// </summary>
+        private static readonly MethodInfo CancelInboundRequestMethodInfo = typeof(StandardCancellationStrategy).GetMethod(nameof(CancelInboundRequest), BindingFlags.Instance | BindingFlags.NonPublic);
+
+        /// <summary>
+        /// A map of id's from inbound calls that have not yet completed and may be canceled,
+        /// to their <see cref="CancellationTokenSource"/> instances.
+        /// </summary>
+        private readonly Dictionary<RequestId, CancellationTokenSource> inboundCancellationSources = new Dictionary<RequestId, CancellationTokenSource>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="StandardCancellationStrategy"/> class.
+        /// </summary>
+        /// <param name="jsonRpc">The <see cref="JsonRpc"/> connection that this strategy is associated with.</param>
+        public StandardCancellationStrategy(JsonRpc jsonRpc)
+        {
+            this.JsonRpc = jsonRpc ?? throw new ArgumentNullException(nameof(jsonRpc));
+
+            // When we add this method, we *must* specify to use a plain SynchronizationContext instance
+            // instead of whatever the user may be using in order to support cancelling of server methods
+            // that have not yet yielded their SyncContext back for calling another method.
+            this.JsonRpc.AddLocalRpcMethod(
+                CancelInboundRequestMethodInfo,
+                this,
+                new JsonRpcMethodAttribute(CancelRequestSpecialMethod),
+                JsonRpc.DefaultSynchronizationContext);
+        }
+
+        /// <summary>
+        /// Gets the <see cref="JsonRpc"/> connection that this strategy is associated with.
+        /// </summary>
+        internal JsonRpc JsonRpc { get; }
+
+        /// <inheritdoc />
+        public virtual void IncomingRequestStarted(RequestId requestId, CancellationTokenSource cancellationTokenSource)
+        {
+            lock (this.inboundCancellationSources)
+            {
+                this.inboundCancellationSources.Add(requestId, cancellationTokenSource);
+            }
+        }
+
+        /// <inheritdoc />
+        public virtual void IncomingRequestEnded(RequestId requestId)
+        {
+            lock (this.inboundCancellationSources)
+            {
+                this.inboundCancellationSources.Remove(requestId);
+            }
+        }
+
+        /// <inheritdoc />
+        public virtual void CancelOutboundRequest(RequestId requestId)
```

Comment:
so IncommingRequestStarted/Ended and AddLocalRpcMethod for CancelInboundRequestMethodInfo is all to handle incoming cancellation message and that seems all good.

for outgoing requests, it seems it only has CancelOutboundRequest but doesn't have a notification for when the outbound request it is canceling has returned. we need that hook to find safe point to remove some state (a file)

Reply:
hmmm.. it looks like (from test and code), IncommingRequestStarted/Ended actually used for out going request as well. then, all good for us!

Reply:
> it looks like (from test and code), IncommingRequestStarted/Ended actually used for out going request as well. then, all good for us!

No, it doesn't. IncomingRequestStarted/Ended only get invoked on the RPC server side. 

>  it seems it only has CancelOutboundRequest but doesn't have a notification for when the outbound request it is canceling has returned. we need that hook to find safe point to remove some state (a file)

This is the kind of review I was hoping to get from you, to make sure you had the invocations you need. I'll add another method so that the RPC client gets a chance to recognize when the request is fullfilled/canceled.

Reply:
I've pushed a change to add the method. But it occurs to me that a race condition may exist where if cancellation is raised at about the same time as the response is received, we may raise the two methods out of order, which would leave a file behind on disk. I'm thinking about how to mitigate that.

Reply:
OK, I added a test and carefully timed it with the debugger to ensure we're good here. The two outbound request-related methods cannot be executed out of order or concurrently, although the `OutboundRequestEnded` method *may* be invoked without an associated `CancelOutboundRequest`. I documented this on the interface and trust folks can deal with it in their implementations. We might be able to avoid that in `JsonRpc` but it would add complexity that I suspect isn't necessary, and it won't happen except in narrow race conditions.